### ISSUE TITLE
Null in Result

### ISSUE THREAD

**Issue by aminaaslam** [jpmml/jpmml-evaluator#52]
Hi villu,
```
ProbabilityDistribution prob = (ProbabilityDistribution) results.get(evaluator.getTargetField().getName());
```
This is returning me null. I dont know whats going on. 
I have tried to match my PMML with the example that you showed me  but even then its failing.
Can you please look into it and guide me.

Thanks

**Comment by vruusmann** [comment]
If one of input values is missing, then the NN model evaluates to a missing result. In that case, the `result` variable itself should be `null`, and `result.get(evaluator.getTargetField().getName())` should fail with an `java.lang.NullPointerException`.

In you case, the `result` variable is non-`null`, which means that the evaluation proceeded to end. You should simply print it to `System.out`, and see what key-value entries are there:
```Java
Map<FieldName, ?> result = evaluator.evaluate(arguments);
System.out.println(result);
```

I suspect there's still something slightly off with your NN data structure. I could take a look at it over the weekend, but you should provide me with some test data - a CSV file with five or so data records would be sufficient.

**Comment by aminaaslam** [comment]
Thanks @vruusmann 
This is the output of `System.out.println(result);`

{Fraud=null}
Fraud

I am attaching sample data. It will be very helpful if u can give some pointers as to what is going on with the evaluation of my PMML.
Thanks
[sampledata.txt](https://github.com/jpmml/jpmml-evaluator/files/852038/sampledata.txt)

**Comment by aminaaslam** [comment]
blockquote, div.yahoo_quoted { margin-left: 0 !important; border-left:1px #715FFA solid !important; padding-left:1ex !important; background-color:white !important; } Hi villu, Did u get a chance to look at my test data and see what is wrong with results? Any help in this regard will be appreciated. Thanks,Amina 


Sent from Yahoo Mail for iPhone


On Friday, March 17, 2017, 6:04 PM, Villu Ruusmann <notifications@github.com> wrote:


If one of input values is missing, then the NN model evaluates to a missing result. In that case, the result variable itself should be null, and result.get(evaluator.getTargetField().getName()) should fail with an java.lang.NullPointerException.

In you case, the result variable is non-null, which means that the evaluation proceeded to end. You should simply print it to System.out, and see what key-value entries are there:
Map<FieldName, ?> result = evaluator.evaluate(arguments);
System.out.println(result);
I suspect there's still something slightly off with your NN data structure. I could take a look at it over the weekend, but you should provide me with some test data - a CSV file with five or so data records would be sufficient.

**Comment by vruusmann** [comment]
Turns out my earlier Evaluator API description (https://github.com/jpmml/jpmml-evaluator/issues/52#issuecomment-287482880) wasn't so accurate after all. Namely, the method `Evaluator#evaluate(Map)` will always return a non-`null` result.

You need to check this `java.util.Map` instance for "interesting" field name-value mappings to understand if the evaluation completed normally or not. Typically, if an evaluation doesn't complete normally, then the target field is mapped to a missing value (represented by `null`). This is exactly the case here.

Your NN data structure appears invalid, because on lines 2141 and 2153 you're referencing a `Neuron` element whose identifier is "0" (eg. `<Con from="0" weight="24.53301996340097"/>`). However, there is no such `Neuron` (or `NeuralInput`) element here. Maybe the indexing of your elements is off by 1?

Currently, JPMML-Evaluator treats "bad" connections as missing input values. However, it should treat them as invalid syntax, and throw an `InvalidFeatureException` instead. This way end users would be able to diagnose and fix the problem more easily.

**Comment by aminaaslam** [comment]
Thanks villu. I will just keep this issue open for alitte bit should i have more related questions.

**Comment by aminaaslam** [comment]
Hi @vruusmann, 
I keep getting the same result even after i fixed the PMML. Here is the updated PMML part
`<NeuralLayer numberOfNeurons="2">
            <Neuron id="131">
                <Con from="0" weight="10.633867526375015"/>
                <Con from="1" weight="-5.681172081646052"/>
                <Con from="2" weight="-27.43669093552691"/>
                <Con from="3" weight="-9.444981138011313"/>
                <Con from="4" weight="-0.20341833471931997"/>
                <Con from="5" weight="-7.995261552034661"/>
                <Con from="6" weight="1.0972652962897753"/>
                <Con from="7" weight="-8.689967742801011"/>
                <Con from="8" weight="-44.66466327283535"/>
                <Con from="9" weight="12.952597701404548"/>
            </Neuron>
            <Neuron id="132">
                <Con from="0" weight="-6.105671494247553"/>
                <Con from="1" weight="-4.725587445097089"/>
                <Con from="2" weight="-5.696613355543679"/>
                <Con from="3" weight="-6.3425717470030865"/>
                <Con from="4" weight="-5.823278634433945"/>
                <Con from="5" weight="-6.122758532943483"/>
                <Con from="6" weight="-2.376126414432059"/>
                <Con from="7" weight="-4.064814864824909"/>
                <Con from="8" weight="-5.736451491181526"/>
                <Con from="9" weight="-5.598418088854917"/>
            </Neuron>
        </NeuralLayer>
        <NeuralLayer numberOfNeurons="2">
            <Neuron id="133" bias="0.0">
                <Con from="132" weight="4.61148100819104"/>
            </Neuron>
            <Neuron id="134" bias="1.0">
                <Con from="132" weight="1.0695827640345894"/>
            </Neuron>
        </NeuralLayer>
        <NeuralOutputs numberOfOutputs="2">
            <NeuralOutput outputNeuron="133">
                <DerivedField displayName="derived_Fraud" optype="continuous" dataType="double">
                    <NormDiscrete field="Fraud" value="0"/>
                </DerivedField>
            </NeuralOutput>
            <NeuralOutput outputNeuron="134">
                <DerivedField displayName="derived_Fraud" optype="continuous" dataType="double">
                    <NormDiscrete field="Fraud" value="1"/>
                </DerivedField>
            </NeuralOutput>
        </NeuralOutputs>`

Any ideas as to what is going on?

**Comment by vruusmann** [comment]
@aminaaslam Your PMML still contains `Con` elements that attempt to reference `Neuron` (or `NeuralInput`) element whose id is `0`:
```
<Con from="0" weight="10.633867526375015"/>
```

I'm expecting to implement some new NN data structure sanity checks into the JPMML-Evaluator library either tomorrow, or the day after that. You may wish to delay your work until that happens.

**Comment by aminaaslam** [comment]
Sorry i  didnt attach the complete file. I fixed the indexes problem.
[PMML.txt](https://github.com/jpmml/jpmml-evaluator/files/859426/PMML.txt)

**Comment by vruusmann** [comment]
At the moment, the only way how a NN model can produce a missing value (ie. `return null`) is when some `NeuralInput` element produced a missing value in the first place:
https://github.com/jpmml/jpmml-evaluator/blob/master/pmml-evaluator/src/main/java/org/jpmml/evaluator/neural_network/NeuralNetworkEvaluator.java#L323

It can only happen when you provided incomplete Map of arguments to the model.

I couldn't use your original sample data file, because it didn't have the header row (column names). I created a temporary header row based on the `/PMML/NeuralNetwork/MiningSchema` element, but it contains one more cell than data rows. As a workaround, I appended a dummy `0` cell to all data rows. See the attached file 
[sampledata_fixed.csv.txt](https://github.com/jpmml/jpmml-evaluator/files/864360/sampledata_fixed.csv.txt)

Your PMML model can score my sample data file without problems (ie. the target value is non-`null`).

**Comment by aminaaslam** [comment]
Thanks @vruusmann for being so prompt. The reason i missed a value in my data row is because Fraud is the target variable so i skipped that and its value. 
Are u telling me that i need to include Target variable's value aswell. Isnt the model suppose to predict???

**Comment by vruusmann** [comment]
In the sample data file, the sixth column contains only `null` values. According to the `/PMML/NeuralNetwork/MiningSchema` element, this is the "Fraud" column? If you say that this sixth column is something else than "Fraud", then the lengths of header row and data rows would match, but you would have missing input values - leading to missing target values.

Try the prediction yourself:
```
$ java -cp pmml-evaluator-example/target/example-1.3-SNAPSHOT.jar org.jpmml.evaluator.EvaluationExample --model PMML.txt --input sampledata_fixed.csv.txt --output nn_fraud.csv
```

Long story short, the NN data structure is now syntactically valid. The prediction fails, because you're executing the model with bad arguments.

**Comment by aminaaslam** [comment]
blockquote, div.yahoo_quoted { margin-left: 0 !important; border-left:1px #715FFA solid !important; padding-left:1ex !important; background-color:white !important; } Thank you so much villu. I am able to execute the model and make predictions off of it. ThanksAmina 


Sent from Yahoo Mail for iPhone


On Thursday, March 23, 2017, 3:58 PM, Villu Ruusmann <notifications@github.com> wrote:


In the sample data file, the sixth column contains only null values. According to the /PMML/NeuralNetwork/MiningSchema element, this is the "Fraud" column? If you say that this sixth column is something else than "Fraud", then the lengths of header row and data rows would match, but you would have missing input values - leading to missing target values.

Try the prediction yourself:
$ java -cp pmml-evaluator-example/target/example-1.3-SNAPSHOT.jar org.jpmml.evaluator.EvaluationExample --model PMML.txt --input sampledata_fixed.csv.txt --output nn_fraud.csv

Long story short, the NN data structure is now syntactically valid. The prediction fails, because you're executing the model with bad arguments.

**Comment by aminaaslam** [comment]
Thanks @vruusmann; i was able to figure out the problem and now the PMML model makes predictions.
